### PR TITLE
Tree hash cache arena

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Proposed Changes

Addresses the excessive memory consumption we were observing whilst running 100k validators. On my desktop this PR demonstrates a reduction of memory consumption from ~8gb to ~2.6gb.

Moves some allocations during cached tree hashing from the heap to the stack showing a ~65% time improvement (credits to @michaelsproul for coming up with this before me).

Adds support for updating the validator registry in parallel, showing at least a 50% improvement on my machine (6-core).

## Additional Info

- ~~Blocked on #384~~

### CODE REVIEW

File: eth2/utils/cached_tree_hash/src/cache.rs [hunk 1]
```diff
@@ -87,24 +127,30 @@ impl TreeHashCache {
...
                         // Parent layer should already contain nodes for all non-dirty indices
-                        if idx != self.layers[depth - 1].len() {
+                        if idx != self.layers[depth - 1].len(arena)? {
                             return Err(Error::CacheInconsistent);
                         }
-                        self.layers[depth - 1].push(Hash256::from_slice(&new_hash));
+                        self.layers[depth - 1]
+                            .push(arena, Hash256::from_slice(&new_hash))
+                            // TODO: fix expect
+                            .expect("should push");
```
Comment (michaelsproul) [inline eth2/utils/cached_tree_hash/src/cache.rs]: Expect

File: eth2/utils/cached_tree_hash/src/cache.rs [hunk 2]
```diff
@@ -113,29 +159,24 @@ impl TreeHashCache {
...
     }
 
     /// Get the root of this cache, without doing any updates/computation.
-    pub fn root(&self) -> Hash256 {
+    pub fn root(&self, arena: &VecArena) -> Hash256 {
         self.layers[0]
-            .get(0)
+            .get(arena, 0)
+            // TODO: deal with expect
+            .expect("arena should be known")
```
Comment (michaelsproul) [inline eth2/utils/cached_tree_hash/src/cache.rs]: Expect

File: eth2/utils/eth2_hashing/src/lib.rs [hunk 3]
```diff
@@ -27,10 +27,18 @@ pub fn hash(input: &[u8]) -> Vec<u8> {
...
+///
+/// Will panic if either `h1` or `h2` are not 32 bytes in length.
+pub fn hash32_concat(h1: &[u8], h2: &[u8]) -> [u8; 32] {
+    let mut preimage = [0; 64];
+    preimage[0..32].copy_from_slice(h1);
+    preimage[32..64].copy_from_slice(h2);
+
+    let mut output = [0; 32];
+    output[..].copy_from_slice(&hash(&preimage));
+    output
```
Comment (michaelsproul) [inline eth2/utils/eth2_hashing/src/lib.rs]: Have you tried [`ring::digest::Context`](https://docs.rs/ring/0.16.9/ring/digest/struct.Context.html) here? It's probably no quicker, but it avoids the panic

File: eth2/utils/cached_tree_hash/src/vec_arena.rs [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+#[derive(Debug, PartialEq, Clone)]
+pub enum Error {
+    UnknownArenaId(usize),
+    OffsetOverflow,
+    OffsetUnderflow,
+    RangeOverFlow,
+}
+
+#[derive(Debug, PartialEq, Clone, Default, Encode, Decode)]
+pub struct VecArena<T: Encode + Decode> {
```
Comment (michaelsproul) [inline eth2/utils/cached_tree_hash/src/vec_arena.rs]: Some comments in this module would be great, at the moment it's got me like :eyes:

File: eth2/utils/eth2_hashing/src/lib.rs (hunk 3 above)
Comment (paulhauner) [reply eth2/utils/eth2_hashing/src/lib.rs]: I tested it, no noticeable improvement but it does avoid a heap allocation so I left it in :)

**Comment by paulhauner** [comment]
Thanks! All comments addressed :)

File: eth2/utils/eth2_hashing/src/lib.rs (hunk 3 above)
Comment (michaelsproul) [reply eth2/utils/eth2_hashing/src/lib.rs]: Cool cool

**Pull request merged**